test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java

```diff
@@ -0,0 +1,387 @@
+            conService.start(true);
+        }
+        conService.scaleService(3, true);
+        List<URI> conUris = conService.getServiceDetails();
+        log.debug("Pravega Controller service  details: {}", conUris);
+
+        // Fetch all the RPC endpoints and construct the client URIs.
+        final List<String> uris = conUris.stream().filter(uri -> uri.getPort() == 9092).map(URI::getAuthority)
+                .collect(Collectors.toList());
+
+        URI controllerURI = URI.create("tcp://" + String.join(",", uris));
+        log.info("Controller Service direct URI: {}", controllerURI);
+
+        //4.start 3 instances of pravega segmentstore
+        Service segService = new PravegaSegmentStoreService("segmentstoreFailover1", zkUri, controllerURI);
+        if (!segService.isRunning()) {
+            segService.start(true);
+        }
+        segService.scaleService(3, true);
+        List<URI> segUris = segService.getServiceDetails();
+        log.debug("Pravega Segmentstore service  details: {}", segUris);
+    }
+
+    @Before
+    public void setup() {
+
+        //Start 1 instance of zookeeper
+        Service zkService = new ZookeeperService("zookeeper");
+        if (!zkService.isRunning()) {
+            zkService.start(true);
+        }
+        List<URI> zkUris = zkService.getServiceDetails();
+        log.debug("Zookeeper service details: {}", zkUris);
+        //get the zk ip details and pass it to  host, controller
+        URI zkUri = zkUris.get(0);
+
+        // Verify controller is running.
+        controllerInstance = new PravegaControllerService("controllerFailover1", zkUri);
+        assertTrue(controllerInstance.isRunning());
+        List<URI> conURIs = controllerInstance.getServiceDetails();
+        log.info("Pravega Controller service instance details: {}", conURIs);
+
+        // Fetch all the RPC endpoints and construct the client URIs.
+        final List<String> uris = conURIs.stream().filter(uri -> uri.getPort() == 9092).map(URI::getAuthority)
+                .collect(Collectors.toList());
+
+        controllerURIDirect = URI.create("tcp://" + String.join(",", uris));
+        log.info("Controller Service direct URI: {}", controllerURIDirect);
+
+        // Verify segment store is running.
+        segmentStoreInstance = new PravegaSegmentStoreService("segmentstoreFailover1", zkUri, controllerURIDirect);
+        assertTrue(segmentStoreInstance.isRunning());
+        log.info("Pravega Segmentstore service instance details: {}", segmentStoreInstance.getServiceDetails());
+    }
+
+    @After
+    public void tearDown() {
+        if (controllerInstance != null && controllerInstance.isRunning()) {
+            controllerInstance.stop();
+        }
+        if (segmentStoreInstance != null && segmentStoreInstance.isRunning()) {
+            segmentStoreInstance.stop();
+        }
+    }
+
+    @Test(timeout = 600000)
+    public void multiReaderWriterWithFailOverTest() throws Exception {
```

Comment:
This test case is good, but I still think we can do better than this. What I'd like to test is:

- Start writers and readers
- At some random point of the execution, crash a segment store server
- At some random point of the execution, crash a controller instance

As is, it is too deterministic and crashes never happen at deterministic points of an execution. Could we change the test case to do this?